src/parse/asp/builtins.go

```diff
@@ -753,6 +754,64 @@ func addDep(s *scope, args []pyObject) pyObject {
 	return None
 }
 
+func addDatumToTargetAndMaybeQueue(s *scope, target *core.BuildTarget, obj pyObject, systemAllowed, tool bool) {
+	datum := core.NewFileLabel(string(obj.(pyString)), s.pkg)
+	target.AddDatum(datum)
+	// Queue this dependency if it'll be needed.
+	if l, ok := datum.Label(); ok && target.State() > core.Inactive {
+		err := s.state.QueueTarget(l, target.Label, true, false)
+		s.Assert(err == nil, "%s", err)
+	}
+}
+
+func addNamedDatumToTargetAndMaybeQueue(s *scope, name string, target *core.BuildTarget, obj pyObject, systemAllowed, tool bool) {
+	datum := core.NewFileLabel(string(obj.(pyString)), s.pkg)
+	target.AddNamedDatum(name, datum)
+	// Queue this dependency if it'll be needed.
+	if l, ok := datum.Label(); ok && target.State() > core.Inactive {
+		err := s.state.QueueTarget(l, target.Label, true, false)
+		s.Assert(err == nil, "%s", err)
+	}
+}
+
+// Add runtime dependencies to target
+func addData(s *scope, args []pyObject) pyObject {
+	s.Assert(s.Callback, "can only be called from a pre- or post-build callback")
+
+	label := args[0]
+	target := getTargetPost(s, string(label.(pyString)))
+
+	systemAllowed := true
+	tool := true
+
+	// add_data() builtin can take a string, list, or dict
+	if isType(args[1], "str") {
+		if bi := ParseBuildInput(s, args[1], string(label.(pyString)), systemAllowed, tool); bi != nil {
+			addDatumToTargetAndMaybeQueue(s, target, args[1], systemAllowed, tool)
+		}
+	} else if isType(args[1], "list") {
+		for _, str := range args[1].(pyList) {
+			if bi := ParseBuildInput(s, str, string(label.(pyString)), systemAllowed, tool); bi != nil {
+				addDatumToTargetAndMaybeQueue(s, target, str, systemAllowed, tool)
+			}
+		}
+	} else if isType(args[1], "dict") {
+		for name, v := range args[1].(pyDict) {
+			for _, str := range v.(pyList) {
+				if bi := ParseBuildInput(s, str, string(label.(pyString)), systemAllowed, tool); bi != nil {
+					addNamedDatumToTargetAndMaybeQueue(s, name, target, str, systemAllowed, tool)
+				}
+			}
+		}
+	} else {
+		log.Error("Unrecognised data type passed to add_data")
```

Comment:
Although we shouldn't hit this line, this should probably be `log.Fatal` otherwise the build will carry on which could lead to some funny situations. `log.Error()` just prints the message to the terminal but doesn't cause the build to terminate.